cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,916 @@
+                       thrust::make_counting_iterator<edge_t>(0),
+                       thrust::make_counting_iterator<edge_t>(intersection_indices.size()),
+                       unroll_edge<vertex_t, edge_t, decltype(transposed_edge_first)>{
+                         edge_t{num_valid_edges},
+                         raft::device_span<edge_t>(num_triangles.data(), num_triangles.size()),
+                         get_dataframe_buffer_begin(vertex_pair_buffer_p_r_edge_p_q),
+                         transposed_edge_first,
+                         transposed_edge_first + num_valid_edges,
+                         transposed_edge_first + edgelist_srcs.size()});
+
+      thrust::for_each(handle.get_thrust_policy(),
+                       thrust::make_counting_iterator<edge_t>(0),
+                       thrust::make_counting_iterator<edge_t>(intersection_indices.size()),
+                       unroll_edge<vertex_t, edge_t, decltype(transposed_edge_first)>{
+                         edge_t{num_valid_edges},
+                         raft::device_span<edge_t>(num_triangles.data(), num_triangles.size()),
+                         get_dataframe_buffer_begin(vertex_pair_buffer_q_r_edge_p_q),
+                         transposed_edge_first,
+                         transposed_edge_first + num_valid_edges,
+                         transposed_edge_first + edgelist_srcs.size()});
+
+      // case 2: unroll (q, r)
+      // For each (q, r) edges to unroll, find the incoming edges to 'r' let's say from 'p' and
+      // create the pair (p, q)
+      cugraph::find_unroll_p_r_or_q_r_edges<vertex_t, edge_t, false, true>(
+        handle,
+        cur_graph_view,
+        num_invalid_edges,
+        num_valid_edges,
+        raft::device_span<vertex_t>(edgelist_srcs.data(), edgelist_srcs.size()),
+        raft::device_span<vertex_t>(edgelist_dsts.data(), edgelist_dsts.size()),
+        raft::device_span<edge_t>(num_triangles.data(), num_triangles.size()));
+
+      // case 3: unroll (p, r)
+      cugraph::find_unroll_p_r_or_q_r_edges<vertex_t, edge_t, false, false>(
+        handle,
+        cur_graph_view,
+        num_invalid_edges,
+        num_valid_edges,
+        raft::device_span<vertex_t>(edgelist_srcs.data(), edgelist_srcs.size()),
+        raft::device_span<vertex_t>(edgelist_dsts.data(), edgelist_dsts.size()),
+        raft::device_span<edge_t>(num_triangles.data(), num_triangles.size()));
+  
+      auto edges_with_triangle_last =
+        thrust::stable_partition(handle.get_thrust_policy(),
+                                 transposed_edge_triangle_count_pair_first,
+                                 transposed_edge_triangle_count_pair_first + num_triangles.size(),
+                                 [] __device__(auto e) {
+                                   auto num_triangles = thrust::get<1>(e);
+                                   return num_triangles > 0;
+                                 });
```

Comment:
Wondering this stable_partition is really necessary. Yes, you need to mask out invalid edges. Is it incorrect leave the  0 triangle count edges in the valid region? They will go to the invalid partition in the beginning of the while loop and will be masked out in the next iteration. I guess the numbers of 0 triangle count edges in the valid region will be small so removing those earlier won't help much in performance. Scanning the entire edge list one more time sounds more expensive than that.

Reply:
And shouldn't be part of the final code but make sure that all the invalid edges have 0 triangle count by now.

Reply:
> Yes, you need to mask out invalid edges

Do we? or do we need to mask out 0 triangle count edges because by the time we reach this stage, all the invalid edges we had at the beginning have 0 triangle count.

Reply:
We are calling `cugraph::transform_e` and filling our edge bucket only once which is at the end of an iteration. I don't see where else we can mask edges before identifying them with `stable_partition` unless I am missing something.